web3.js/src/loader.js

```diff
@@ -45,36 +45,87 @@ export class Loader {
    * @param program Account to load the program into
    * @param programId Public key that identifies the loader
    * @param data Program octets
+   * @return true if program was loaded successfully, false if program was already loaded
    */
   static async load(
     connection: Connection,
     payer: Account,
     program: Account,
     programId: PublicKey,
     data: Buffer | Uint8Array | Array<number>,
-  ): Promise<void> {
+  ): Promise<boolean> {
     {
       const balanceNeeded = await connection.getMinimumBalanceForRentExemption(
         data.length,
       );
-      const transaction = new Transaction().add(
-        SystemProgram.createAccount({
-          fromPubkey: payer.publicKey,
-          newAccountPubkey: program.publicKey,
-          lamports: balanceNeeded > 0 ? balanceNeeded : 1,
-          space: data.length,
-          programId,
-        }),
-      );
-      await sendAndConfirmTransaction(
-        connection,
-        transaction,
-        [payer, program],
-        {
-          commitment: 'single',
-          skipPreflight: true,
-        },
+
+      // Fetch program account info to check if it has already been created
+      const programInfo = await connection.getAccountInfo(
+        program.publicKey,
+        'single',
       );
+
+      let transaction: Transaction | null = null;
+      if (programInfo !== null) {
+        if (programInfo.executable) {
+          return false;
+        }
+
+        if (programInfo.data.length !== data.length) {
+          transaction = transaction || new Transaction();
+          transaction.add(
+            SystemProgram.allocate({
+              accountPubkey: program.publicKey,
+              space: data.length,
+            }),
+          );
+        }
+
+        if (!programInfo.owner.equals(programId)) {
+          transaction = transaction || new Transaction();
+          transaction.add(
+            SystemProgram.assign({
+              accountPubkey: program.publicKey,
+              programId,
+            }),
+          );
+        }
+
+        if (programInfo.lamports < balanceNeeded) {
+          transaction = transaction || new Transaction();
+          transaction.add(
+            SystemProgram.transfer({
+              fromPubkey: payer.publicKey,
+              toPubkey: program.publicKey,
+              lamports: balanceNeeded - programInfo.lamports,
+            }),
+          );
+        }
+      } else {
```

Comment:
Another option here is if the account is created correctly but not executable it probably means it was not finalized, we could go through and look for the gaps and fill them in.  We currently just go and rewrite the all the data which is ok but wasteful

Reply:
It doesn't prevent success but it is pretty wasteful. Are you cool with that being done in a separate PR?

Reply:
https://github.com/solana-labs/solana/issues/12261